scripts/compile_commits.sh

```diff
@@ -0,0 +1,224 @@
+# It Automatically uses the latest supported version of GCC and Clang,
+# building with both compilers if available.
+#
+# When a failure occurs, the log is printed and the problematic commit
+# will be checked out, so you can inspect and amend the commit.
+#
+# CLI tools used:
+# - BASH 4.4+, for readarray
+# - GNU ls, for natural-sorting the native files by version
+# - GNU sed, to parse the default branch
+# - Git
+# - All the usual build depenencies (Meson, Ninja, etc)
+#
+
+set -euo pipefail
+
+# Common variables that are setup and used below
+declare temp_file
+declare head_commit
+declare -a commits
+declare -a compilers
+declare meson_setup
+
+# GNU tools
+declare LS
+declare SED
+
+populate_gnu_ls() {
+	# shellcheck disable=2010
+	if \ls --version 2>/dev/null | grep -q GNU; then
+		LS=$(command -v ls)
+
+	elif command -v gdir >/dev/null; then
+		LS=$(command -v gdir)
+
+	else
+		echo "Please install corutils with your package manager and then:"
+		echo " - Adjust your PATH to find GNU 'ls' first, or"
+		echo " - Symlink 'ls' to your GNU 'ls' binary, or"
+		echo " - If on macOS, simply 'brew install coreutils' and you're done!"
+		exit 1
+	fi
+}
+
+populate_gnu_sed() {
+	if \sed --version 2>/dev/null | grep -q GNU; then
+		SED=$(command -v sed)
+
+	elif command -v gsed >/dev/null; then
+		SED=$(command -v gsed)
+
+	else
+		echo "Please install GNU sed with your package manager."
+		echo " - If on macOS, simply 'brew install gnu-sed' and you're done!"
+		exit 1
+	fi
+}
+
+extract_cpp_from_native_file() {
+	grep cpp "$1" | cut -d "'" -f 4
+}
+
+get_native_file_for_compiler() {
+	local compiler
+	compiler="$1"
+
+	populate_gnu_ls
+
+	local native_file
+	for native_file in $("$LS" \
+	                     --reverse \
+	                     --sort=version \
+	                     .github/meson/native-"$compiler"*.ini); do
+
+		local cpp
+		cpp=$(extract_cpp_from_native_file "$native_file")
+		if command -v "$cpp" &>/dev/null; then
+			echo "$native_file"
+			return
+		fi
+	done
+}
+
+allocate_temp_file() {
+	temp_file=$(mktemp 2>/dev/null || mktemp -t temp)
+	echo "Realtime log: $temp_file"
+
+	# Cleanup the log on termination
+	trap 'rm -f "$temp_file"' EXIT
+}
+
+run() {
+	# Split the command into an array
+	local -a cmds
+	IFS=" " read -r -a cmds <<<"$1"
+
+	# Execute the command with arguments
+	if ! "${cmds[@]}" &>"$temp_file"; then
+		echo "failed"
+		echo ""
+		echo "---"
+		cat "$temp_file"
+		echo "---"
+		echo ""
+		return 1
+	fi
+	return 0
+}
+
+populate_commits() {
+
+	populate_gnu_sed
+
+	local default_branch
+	default_branch=$(git symbolic-ref refs/remotes/origin/HEAD 2>/dev/null |
+	                 "$SED" 's@^refs/remotes/origin/@@')
+
+	if [[ -z $default_branch ]]; then
+		echo "Failed to determine the default branch."
+		exit 1
+	fi
+	readarray -t commits < <(git log \
+	                         --format="%h" HEAD \
+	                         --not "$default_branch")
+
```

Comment:
`readarray` is not supported under Zsh on macOS nor, I believe, on older versions of Bash on Mac. Consider a loop for portability, e.g.:

```shell
commits=()
	while IFS= read -r line; do
		commits+=("$line")
	done < <(git log --format="%h" HEAD --not "$default_branch")
```

Reply:
Thanks for that fix up, @kklobe !

Reply:
Nice - you've got GCC 13.x building too? 
I need to install that and get it in my path on the M1 mini :-)

Reply:
yep, the GCC formulae on homebrew have come a long ways in the past year, I need it primarily for gfortan with R packages.

Reply:
GCC 13.0.0 working great here now on the M1, thanks @kklobe :+1: 

```text
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: arm64-apple-darwin20.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```